### PR TITLE
Add attributes and parent pointers in the HTML chapter

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR adds the `attributes` and `parent` properties to `ElementNode`s, and `parent` and (empty) `children` properties to `TextNode`s. The `ElementNode` changes are required by later chapters, so it makes sense to do them here, and they also clarify the stack handling in the parser just a little bit.

### CODE REVIEW

File: src/lab4.py [hunk 1]
```diff
@@ -122,19 +126,25 @@ def parse(tokens):
         implicit_tags(tok, currently_open)
         if isinstance(tok, Text):
             if tok.text.isspace(): continue
-            node = TextNode(tok.text)
-            currently_open[-1].children.append(node)
+            parent = currently_open[-1]
```
Comment (chrishtr) [inline src/lab4.py:129]: You could factor this ilne out of the containing if statement:

if currently_open:
  parent = currently_open[-1]
else
  parent = None

and remove lines 129, 137 and 143-146.

File: src/lab5.py [hunk 2]
```diff
@@ -121,20 +125,26 @@ def parse(tokens):
     for tok in tokens:
         implicit_tags(tok, currently_open)
         if isinstance(tok, Text):
-            node = TextNode(tok.text)
-            if not currently_open: continue
-            currently_open[-1].children.append(node)
+            if tok.text.isspace(): continue
+            parent = currently_open[-1]
```
Comment (chrishtr) [inline src/lab5.py:129]: Same here - refactor?

File: src/lab6.py [hunk 3]
```diff
@@ -129,20 +131,26 @@ def parse(tokens):
     for tok in tokens:
         implicit_tags(tok, currently_open)
         if isinstance(tok, Text):
-            node = TextNode(tok.text)
-            if not currently_open: continue
-            currently_open[-1].children.append(node)
+            if tok.text.isspace(): continue
+            parent = currently_open[-1]
```
Comment (chrishtr) [inline src/lab6.py:135]: ..

**Review by chrishtr — APPROVED** [review]
Approved with some suggestions.

**Comment by pavpanchekha** [comment]
Thanks, great suggestions. Will merge once everything passes.

**Pull request merged**